File: app/src/main/java/com/developer/valyutaapp/ui/main/MainPresenter.java
package com.developer.valyutaapp.ui.main;

import android.content.Context;
import android.util.Log;
import android.widget.Toast;

import com.developer.valyutaapp.data.local_data_source.repository.ValuteRespository;
import com.developer.valyutaapp.data.remote_data_source.NetworkClient;
import com.developer.valyutaapp.data.remote_data_source.NetworkInterfaces;
import com.developer.valyutaapp.model.ValCurs;
import com.developer.valyutaapp.model.Valute;
import com.developer.valyutaapp.utils.Utils;

import java.util.List;

import io.reactivex.Observable;
import io.reactivex.ObservableEmitter;
import io.reactivex.ObservableOnSubscribe;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.annotations.NonNull;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Action;
import io.reactivex.functions.Consumer;
import io.reactivex.observers.DisposableObserver;
import io.reactivex.schedulers.Schedulers;

public class MainPresenter implements MainPresenterInterface {

    MainViewInterface mvi;
    Context context;
    private String TAG = "MainPresenter";
    private ValuteRespository valuteRespository;
    private CompositeDisposable compositeDisposable;
    private int countVal = 0;

    public MainPresenter(Context context, MainViewInterface mvi, ValuteRespository valuteRespository) {
        this.context = context;
        this.mvi = mvi;
        this.valuteRespository = valuteRespository;
        compositeDisposable = new CompositeDisposable();
    }

    @Override
    public void getValutesRemote() {
        mvi.showProgressBar();
        loadValuteCount();
    }

    @Override
    public void getValutesLocal() {
        loadData();
    }

    public Observable<ValCurs> getObservable(){
        return NetworkClient.getRetrofit().create(NetworkInterfaces.class)
                .getValutes(Utils.getDate(),"xmlout")
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread());
    }

    public DisposableObserver<ValCurs> getObserver(int valCount){
        return new DisposableObserver<ValCurs>() {

            @Override
            public void onNext(@NonNull ValCurs valCurs) {
                Log.d(TAG,"OnNext"+valCurs.getValute().toString());
                if (valCount > 0){
                    updateData(valCurs);
                }else {
                    insertData(valCurs);
                }
            }

            @Override
            public void onError(@NonNull Throwable e) {
                Log.d(TAG,"Error"+e);
                e.printStackTrace();
                mvi.displayError("Error fetching valutes Data");
            }

            @Override
            public void onComplete() {
                Log.d(TAG,"Completed");
                mvi.hideProgressBar();
            }
        };
    }

    public void insertData(ValCurs valCurs){
        Disposable disposable = io.reactivex.Observable.create(new ObservableOnSubscribe<Object>() {
            @Override
            public void subscribe(ObservableEmitter<Object> e) {
                Log.d(TAG,"subscribe = " + valCurs.valute.size());
                valuteRespository.insertValute(valCurs.valute);
                e.onComplete();
            }
        })
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribe(new Consumer() {
                    @Override
                    public void accept(Object o) {
                        Toast.makeText(context, "Valute added!", Toast.LENGTH_SHORT).show();
                    }
                }, new Consumer<Throwable>() {
                    @Override
                    public void accept(Throwable throwable) {
                        Toast.makeText(context, ""+ throwable.getMessage(), Toast.LENGTH_SHORT).show();
                    }
                }, new Action(){
                    @Override
                    public void run() throws Exception {
                        loadData();
                    }
                });
    }


    public void updateData(ValCurs valCurs){
        Disposable disposable = io.reactivex.Observable.create(new ObservableOnSubscribe<Object>() {
            @Override
            public void subscribe(ObservableEmitter<Object> e) {
                Log.d(TAG,"subscribe = " + valCurs.valute.size());
                valuteRespository.updateValute(valCurs.valute);
                e.onComplete();
            }
        })
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribe(new Consumer() {
                    @Override
                    public void accept(Object o) {
                        Toast.makeText(context, "Valute added!", Toast.LENGTH_SHORT).show();
                    }
                }, new Consumer<Throwable>() {
                    @Override
                    public void accept(Throwable throwable) {
                        Toast.makeText(context, ""+ throwable.getMessage(), Toast.LENGTH_SHORT).show();
                    }
                }, new Action(){
                    @Override
                    public void run() throws Exception {
                        loadData();
                    }
                });
    }


    private int loadValuteCount() {
        Disposable disposable = valuteRespository.getValuteCount()
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribe(new Consumer<Integer>(){
                    @Override
                    public void accept(Integer count) throws Exception {
                        getObservable().subscribeWith(getObserver(count));
                    }
                }, new Consumer<Throwable>() {
                    @Override
                    public void accept(Throwable throwable) throws Exception{
                        Toast.makeText(context, ""+ throwable.getMessage(), Toast.LENGTH_SHORT).show();
                    }
                });
        compositeDisposable.add(disposable);
        return countVal;
    }

    private void loadData() {
        Disposable disposable = valuteRespository.getAllValutes()
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribe(new Consumer<List<Valute>>(){
                    @Override
                    public void accept(List<Valute> valutes) throws Exception {
                        Log.d(TAG,"Completed = " + valutes.size());
                        mvi.displayValutes(valutes);
                        mvi.hideProgressBar();
                    }
                }, new Consumer<Throwable>() {
                    @Override
                    public void accept(Throwable throwable) throws Exception{
                        Toast.makeText(context, ""+ throwable.getMessage(), Toast.LENGTH_SHORT).show();
                    }
                });
        compositeDisposable.add(disposable);
    }
}

File: app/src/main/java/com/developer/valyutaapp/ui/main/MainPresenter.kt
//package com.developer.valyutaapp.ui.main
//
//import android.content.Context
//import android.util.Log
//import io.reactivex.disposables.CompositeDisposable
//import com.developer.valyutaapp.domain.entities.ValCurs
//import io.reactivex.schedulers.Schedulers
//import io.reactivex.android.schedulers.AndroidSchedulers
//import io.reactivex.observers.DisposableObserver
//import android.widget.Toast
//import com.developer.valyutaapp.utils.Utils
//import io.reactivex.Observable
//
//class MainPresenter(
//    var context: Context,
//    var mvi: MainViewInterface,
//    private val valuteRespository: ValuteRespository
//) : MainPresenterInterface {
//
//    private val TAG = "MainPresenter"
//    private val compositeDisposable: CompositeDisposable = CompositeDisposable()
//    private val countVal = 0
//    override fun valutesRemote() {
//        mvi.showProgressBar()
//        loadValuteCount()
//    }
//
//    override fun valutesLocal() {
//        loadData()
//    }
//
//    val observable: Observable<ValCurs>
//        get() = NetworkClient.getRetrofit().create(
//            NetworkInterfaces::class.java
//        )
//            .getValutes(Utils.date, "xmlout")
//            .subscribeOn(Schedulers.io())
//            .observeOn(AndroidSchedulers.mainThread())
//
//    fun getObserver(valCount: Int): DisposableObserver<ValCurs?> {
//        return object : DisposableObserver<ValCurs?>() {
//            override fun onNext(valCurs: ValCurs) {
//                //Log.d(TAG, "OnNext" + valCurs.getValute().toString())
//                if (valCount > 0) {
//                    updateData(valCurs)
//                } else {
//                    insertData(valCurs)
//                }
//            }
//            override fun onError(e: Throwable) {
//                Log.d(TAG, "Error$e")
//                e.printStackTrace()
//                mvi.displayError("Error fetching valutes Data")
//            }
//
//            override fun onComplete() {
//                Log.d(TAG, "Completed")
//                mvi.hideProgressBar()
//            }
//        }
//    }
//
//    fun insertData(valCurs: ValCurs) {
//        val disposable = Observable.create<Any> { e ->
//            Log.d(TAG, "subscribe = " + valCurs.valute!!.size)
//            valuteRespository.insertValute(valCurs.valute)
//            e.onComplete()
//        }
//            .observeOn(AndroidSchedulers.mainThread())
//            .subscribeOn(Schedulers.io())
//            .subscribe(
//                { Toast.makeText(context, "Valute added!", Toast.LENGTH_SHORT).show() },
//                { throwable ->
//                    Toast.makeText(context, "" + throwable.message, Toast.LENGTH_SHORT).show()
//                }) { loadData() }
//    }
//
//    fun updateData(valCurs: ValCurs) {
//        val disposable = Observable.create<Any> { e ->
//            Log.d(TAG, "subscribe = " + valCurs.valute!!.size)
//            valuteRespository.updateValute(valCurs.valute)
//            e.onComplete()
//        }
//            .observeOn(AndroidSchedulers.mainThread())
//            .subscribeOn(Schedulers.io())
//            .subscribe(
//                { Toast.makeText(context, "Valute added!", Toast.LENGTH_SHORT).show() },
//                { throwable ->
//                    Toast.makeText(context, "" + throwable.message, Toast.LENGTH_SHORT).show()
//                }) { loadData() }
//    }
//
//    private fun loadValuteCount(): Int {
//        val disposable = valuteRespository.valuteCount
//            .observeOn(AndroidSchedulers.mainThread())
//            .subscribeOn(Schedulers.io())
//            .subscribe({ count -> observable.subscribeWith(getObserver(count)) }) { throwable ->
//                Toast.makeText(
//                    context,
//                    "" + throwable.message,
//                    Toast.LENGTH_SHORT
//                ).show()
//            }
//        compositeDisposable.add(disposable)
//        return countVal
//    }
//
//    private fun loadData() {
//        val disposable = valuteRespository.allValutes
//            .observeOn(AndroidSchedulers.mainThread())
//            .subscribeOn(Schedulers.io())
//            .subscribe({ valutes ->
//                Log.d(TAG, "Completed = " + valutes.size)
//                mvi.displayValutes(valutes)
//                mvi.hideProgressBar()
//            }) { throwable ->
//                Toast.makeText(context, "" + throwable.message, Toast.LENGTH_SHORT).show()
//            }
//        compositeDisposable.add(disposable)
//    }
//
//}